File: packages/vpn-routing/src/vpn_routing/node_registry.py
from __future__ import annotations

import json
from dataclasses import asdict
from pathlib import Path

from .models import HealthMetrics, NodeDescriptor, NodeRole


class NodeRegistry:
    def __init__(self, nodes: list[NodeDescriptor]) -> None:
        self._nodes = tuple(nodes)
        self._index = {node.node_id: node for node in nodes}
        if len(self._nodes) != len(self._index):
            raise ValueError("Duplicate node_id detected in registry.")

    @classmethod
    def from_dict(cls, payload: dict[str, object]) -> "NodeRegistry":
        raw_nodes = payload.get("nodes")
        if not isinstance(raw_nodes, list):
            raise ValueError("Registry payload must contain list field 'nodes'.")

        nodes: list[NodeDescriptor] = []
        for raw in raw_nodes:
            if not isinstance(raw, dict):
                continue
            node_id = str(raw.get("node_id", "")).strip()
            if not node_id:
                raise ValueError("node_id is required for each node.")
            role = _coerce_role(raw.get("role"))
            region = str(raw.get("region", "")).strip() or "unknown"
            domain = str(raw.get("domain", "")).strip()
            port = int(raw.get("port", 0))
            transport = str(raw.get("transport", "tcp")).strip() or "tcp"
            active = bool(raw.get("active", True))

            health = None
            raw_health = raw.get("health")
            if isinstance(raw_health, dict):
                health = HealthMetrics(
                    latency_ms=float(raw_health.get("latency_ms", 0.0)),
                    error_rate=float(raw_health.get("error_rate", 0.0)),
                    uptime_ratio=float(raw_health.get("uptime_ratio", 1.0)),
                    load_ratio=float(raw_health.get("load_ratio", 0.0)),
                    last_updated_utc=str(raw_health.get("last_updated_utc", "")),
                )

            nodes.append(
                NodeDescriptor(
                    node_id=node_id,
                    role=role,
                    region=region,
                    domain=domain,
                    port=port,
                    transport=transport,
                    active=active,
                    health=health,
                )
            )
        return cls(nodes=nodes)

    @classmethod
    def from_json_file(cls, path: str | Path) -> "NodeRegistry":
        payload = json.loads(Path(path).read_text(encoding="utf-8"))
        if not isinstance(payload, dict):
            raise ValueError("Registry JSON root must be an object.")
        return cls.from_dict(payload)

    def to_dict(self) -> dict[str, object]:
        return {"nodes": [asdict(node) for node in self._nodes]}

    def list_nodes(self) -> tuple[NodeDescriptor, ...]:
        return self._nodes

    def get_node(self, node_id: str) -> NodeDescriptor | None:
        return self._index.get(node_id)

    def list_active(self, role: NodeRole | None = None, region: str | None = None) -> tuple[NodeDescriptor, ...]:
        out: list[NodeDescriptor] = []
        for node in self._nodes:
            if not node.active:
                continue
            if role is not None and node.role != role:
                continue
            if region is not None and node.region != region:
                continue
            out.append(node)
        return tuple(out)


def _coerce_role(raw: object) -> NodeRole:
    value = str(raw or "backup").lower().strip()
    if value in {"main", "rf", "backup"}:
        return value  # type: ignore[return-value]
    return "backup"

File: packages/vpn-routing/src/node-registry.ts
import { readFileSync } from "node:fs";
import type { NodeDescriptor, NodeRole } from "./types.js";

const VALID_ROLES: ReadonlySet<NodeRole> = new Set(["main", "rf", "backup"]);

export class NodeRegistry {
  private readonly nodes: NodeDescriptor[];
  private readonly index: Map<string, NodeDescriptor>;

  constructor(nodes: NodeDescriptor[]) {
    this.nodes = nodes;
    this.index = new Map(nodes.map((node) => [node.nodeId, node]));
    if (this.index.size !== this.nodes.length) {
      throw new Error("Duplicate nodeId detected in registry.");
    }
  }

  static fromObject(payload: unknown): NodeRegistry {
    if (!payload || typeof payload !== "object") {
      throw new Error("Registry payload must be an object.");
    }

    const maybeNodes = (payload as { nodes?: unknown }).nodes;
    if (!Array.isArray(maybeNodes)) {
      throw new Error("Registry payload must contain array field 'nodes'.");
    }

    const nodes = maybeNodes
      .map((raw) => NodeRegistry.parseNode(raw))
      .filter((node): node is NodeDescriptor => node !== null);

    return new NodeRegistry(nodes);
  }

  static fromJsonFile(path: string): NodeRegistry {
    const content = readFileSync(path, "utf8");
    return NodeRegistry.fromObject(JSON.parse(content) as unknown);
  }

  private static parseNode(raw: unknown): NodeDescriptor | null {
    if (!raw || typeof raw !== "object") {
      return null;
    }
    const item = raw as Record<string, unknown>;
    const nodeId = String(item.node_id ?? item.nodeId ?? "").trim();
    if (!nodeId) {
      throw new Error("nodeId/node_id is required for each node.");
    }

    const roleRaw = String(item.role ?? "backup").trim().toLowerCase();
    const role = VALID_ROLES.has(roleRaw as NodeRole) ? (roleRaw as NodeRole) : "backup";
    const region = String(item.region ?? "unknown").trim() || "unknown";
    const domain = String(item.domain ?? "").trim();
    const port = Number(item.port ?? 0);
    const transport = String(item.transport ?? "tcp").trim() || "tcp";
    const active = item.active === undefined ? true : Boolean(item.active);

    const rawHealth = item.health;
    let health: NodeDescriptor["health"];
    if (rawHealth && typeof rawHealth === "object") {
      const h = rawHealth as Record<string, unknown>;
      health = {
        latencyMs: Number(h.latency_ms ?? h.latencyMs ?? 0),
        errorRate: Number(h.error_rate ?? h.errorRate ?? 0),
        uptimeRatio: Number(h.uptime_ratio ?? h.uptimeRatio ?? 1),
        loadRatio: Number(h.load_ratio ?? h.loadRatio ?? 0),
        lastUpdatedUtc: String(h.last_updated_utc ?? h.lastUpdatedUtc ?? "")
      };
    }

    return {
      nodeId,
      role,
      region,
      domain,
      port,
      transport,
      active,
      ...(health !== undefined ? { health } : {})
    };
  }

  toObject(): { nodes: Array<Record<string, unknown>> } {
    return {
      nodes: this.nodes.map((node) => ({
        node_id: node.nodeId,
        role: node.role,
        region: node.region,
        domain: node.domain,
        port: node.port,
        transport: node.transport,
        active: node.active,
        health: node.health
          ? {
              latency_ms: node.health.latencyMs,
              error_rate: node.health.errorRate,
              uptime_ratio: node.health.uptimeRatio,
              load_ratio: node.health.loadRatio,
              last_updated_utc: node.health.lastUpdatedUtc
            }
          : undefined
      }))
    };
  }

  listNodes(): NodeDescriptor[] {
    return [...this.nodes];
  }

  getNode(nodeId: string): NodeDescriptor | undefined {
    return this.index.get(nodeId);
  }

  listActive(filters?: { role?: NodeRole; region?: string }): NodeDescriptor[] {
    return this.nodes.filter((node) => {
      if (!node.active) {
        return false;
      }
      if (filters?.role && node.role !== filters.role) {
        return false;
      }
      if (filters?.region && node.region !== filters.region) {
        return false;
      }
      return true;
    });
  }
}
